src/main.c

```diff
@@ -26,94 +44,93 @@ install_snaps_cb(GObject *object, GAsyncResult *result, gpointer user_data)
     }
 
-    if (missing_snaps->len == 0) {
+    state->gtk_theme_status = GPOINTER_TO_INT (g_hash_table_lookup (gtk_theme_status, state->gtk_theme_name));
+    state->icon_theme_status = GPOINTER_TO_INT (g_hash_table_lookup (icon_theme_status, state->icon_theme_name));
+    state->sound_theme_status = GPOINTER_TO_INT (g_hash_table_lookup (sound_theme_status, state->sound_theme_name));
+
+    gboolean themes_available = state->gtk_theme_status == SNAPD_THEME_STATUS_AVAILABLE ||
+                                state->icon_theme_status == SNAPD_THEME_STATUS_AVAILABLE ||
+                                state->sound_theme_status == SNAPD_THEME_STATUS_AVAILABLE;
+
+    if (!themes_available) {
         g_print("All available theme snaps installed\n");
         return;
     }
-    g_print("Missing theme snaps:\n");
-    for (i = 0; i < missing_snaps->len; i++) {
-        SnapdSnap *snap = missing_snaps->pdata[i];
 
-        const char* snap_name = snapd_snap_get_name(snap);
-        g_print(" - %s\n", snap_name);
-    }
+    g_print("Missing theme snaps\n");
 
     NotifyNotification *notification = notify_notification_new("Some required theme snaps are missing.", "Would you like to install them now?", "dialog-question");
-
-    g_autoptr(install_info_t) find_data = g_new0(install_info_t, 1);
-    find_data->helper = g_object_ref(helper);
-    find_data->missing_snaps = g_ptr_array_ref(missing_snaps);
-
-    notify_notification_add_action(g_object_ref(notification), "yes", "Yes", install_snaps, g_steal_pointer(&find_data), NULL);
-    notify_notification_add_action(g_object_ref(notification), "no", "No", install_snaps, NULL, NULL);
+    notify_notification_add_action(g_object_ref(notification), "yes", "Yes", notify_cb, state, NULL);
+    notify_notification_add_action(g_object_ref(notification), "no", "No", notify_cb, state, NULL);
 
     notify_notification_show(notification, NULL);
     g_object_unref(notification);
 }
 
 static void
-theme_changed(DsThemeWatcher *watcher, const DsThemeSet *themes, DsSnapdHelper *snapd)
+theme_changed_cb(DsState *state, const DsThemeSet *themes)
 {
     g_message("New theme: gtk=%s icon=%s cursor=%s, sound=%s",
               themes->gtk_theme_name,
               themes->icon_theme_name,
               themes->cursor_theme_name,
               themes->sound_theme_name);
 
-    ds_snapd_helper_find_missing_snaps(snapd, themes, NULL, missing_snaps_ready, NULL);
+    g_free (state->gtk_theme_name);
+    state->gtk_theme_name = g_strdup(themes->gtk_theme_name);
+    state->gtk_theme_status = 0;
+
+    g_free (state->icon_theme_name);
+    state->icon_theme_name = g_strdup(themes->icon_theme_name);
+    state->icon_theme_status = 0;
+
+    g_free (state->sound_theme_name);
+    state->sound_theme_name = g_strdup(themes->sound_theme_name);
+    state->sound_theme_status = 0;
+
+    gchar *gtk_theme_names[2] = { themes->gtk_theme_name, NULL };
+    gchar *icon_theme_names[2] = { themes->icon_theme_name, NULL };
```

Comment:
`cursor_theme_name` should also be included in the list of icon themes to check for (as cursors are also looked up according to the icon theme spec).